### ISSUE TITLE
Does not like database names with spaces

### ISSUE THREAD

**Issue by CoryAlbrecht** [bstroebl/DataDrivenInputMask#13]
I get the following error for "Initialize Layer" and for "Show Input Form" (when object selected):

`DdError: Could not connect to PostgreSQL database:Alternate`

The full name of the database is "Alternate History".

Nothing happens for "Configure Mask"

Using 2.18.24 with GRASS 7 on Windows 10.

**Comment by bstroebl** [comment]
The plugin determines the database credentials of a layer by analyzing its source string. A typical source string is e.g.
`u'dbname=\'mydb\' host=myhosturl port=5432 user=\'myuser\' password=\'mypassword\' sslmode=require key=\'mykey\' srid=4326 type=LineString table="myschema"."mytable" (the_geom) sql='`
This source string is then split at all spaces to get its individual members. Now if the database name contains a space all after it is ignored, resulting in the error you get.
Long story short: there is nothing I can do about it, except to advise you to not use spaces in database names, table names, field names, nor any non-ascii characters.

**Comment by CoryAlbrecht** [comment]
Except spaces are allowed by QGIS, you just need a better parsing algorithm. :-) Perhaps a regex with named capture groups instead of simply splitting by spaces?

```
>>> l = iface.activeLayer().dataProvider().dataSourceUri()
u'dbname=\'Alternate History\' host=pg0.db.cory.albrecht.name port=5432 user=\'maps\' password=\'r&67$7h0HAD!cltc\' key=\'feature_uuid\' srid=4326 type=MultiPolygon table="Balance of Power"."timeline-countries" (geom) sql="start_date" < \'1849/01/14\' AND "end_date" >= \'1849/01/13\' '
>>> p = re.compile('(dbname=\'(?P<value>[^\']*)\' *)+')
>>> m = p.match(l)
>>> m.groups()
(u"dbname='Alternate History' ", u'Alternate History')
>>> m.group('value')
u'Alternate History'
```

**Comment by bstroebl** [comment]
Will try your code, I am no regex expert at all
Still I would avoid spaces; same in Windows: spaces are allowed in paths, Windows can handle it but all the time there are programs that run into problems. So better use an underscore, it makes life easier :-)

**Comment by bstroebl** [comment]
new versions for QGIS 2 and 3 are available
